File: 1249-MinimumRemoveToMakeValidParentheses.cpp
class Solution {
public:
    using pi = pair<char, int>;
    string minRemoveToMakeValid(string str) {
        stack<pi> s;
        for(int i = 0; i<str.length(); i++){
            switch (str[i]){
                case '(':
                    s.push(make_pair(str[i], i));
                    break;
                case ')':
                    if(s.empty()){
                        s.push(make_pair(str[i], i));
                    } else {
                        if(s.top().first=='('){
                            s.pop();
                        }else {
                            s.push(make_pair(str[i], i));
                        }
                    }
                    break;
                default:
                    break;
            }
        }
        if(s.empty()){
            return str;
        } else {
            string res = str;
            while(!s.empty()){
                auto tmp = res;
                tmp.erase(s.top().second, 1);
                s.pop();
                res = tmp;
            }
            return res;
        }
        return "";
    }
};

File: 1249-MinimumRemoveToMakeValidParentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        char[] arr;
        // allocating memory for 5 integers.
        int n = s.length();
        arr = new char[n];
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
 
            // Check if the character has paranthesis
            if((s.charAt(i) == '(') || (s.charAt(i) == ')'))
            {
                if(s.charAt(i)=='(')
                {
                    stack.push(i);
                }
                if(s.charAt(i)==')')
                {
                    if(!stack.empty())
                    {
                        int g = stack.pop();
                        arr[g] = '(';
                        arr[i] = ')';
                    }
                }
            } 
            else
            {
                arr[i]= s.charAt(i);
                
            }   
        }
    StringBuilder str = new StringBuilder();

    for(int i=0;i< arr.length;i++)
    {
        if(arr[i]== '\0')
        {}
        else
        {
            str.append(arr[i]);
        }
    }
    return str.toString();
    }
}
